File: bench/fib/fib.ex
defmodule Fib do
  def fib(0) do 0 end
  def fib(1) do 1 end
  def fib(n) do fib(n-1) + fib(n-2) end

  def fib2(n, a \\ 0, b \\ 1)
  def fib2(0, a, _b) do a end
  def fib2(n, a, b) do fib2(n - 1, b, a + b) end
end

timeit = fn({m, f, a}) ->
  start = :os.system_time(:seconds)
  res = apply(m, f, a)
  delta = :os.system_time(:seconds) - start
  {delta, res}
end

for n <- [10, 25, 35, 100, 100000, 150000, 150500, 1505000] do
  IO.puts "Fib of #{n}"
  {time, value} = timeit.({Fib, :fib2, [n]})
  IO.puts "    time: #{time}"
end

File: bench/fib/fib.lua
-- https://gist.githubusercontent.com/SegFaultAX/2772595/raw/095fd6107259e62eac107e5aa94bd6c7c6508101/fib.lua
function tail_call(n)
  local function inner(m, a, b)
    if m == 0 then
      return a
    end
    return inner(m-1, b, a+b)
  end
  return inner(n, 0, 1)
end

function timeit(f, ...)
  local start = os.time()
  local res = { f(...) }
  local delta = os.time() - start
  return delta, unpack(res)
end

for _, n in ipairs({10, 25, 35, 100}) do
  print("Fib of "..n)
  print(string.format("    time: %s value: %s", timeit(tail_call, n)))
end
